src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1303,17 +1302,17 @@ private void InheritNullableStateOfMember(int targetContainerSlot, int valueCont
 
         /// <summary>
         /// Whenever assigning a variable, and that variable is not declared at the point the state is being set,
-        /// and the new state might be <see cref="NullableFlowState.MaybeNull"/>, this method should be called to perform the
+        /// and the new state is not <see cref="NullableFlowState.NotNull"/>, this method should be called to perform the
         /// state setting and to ensure the mutation is visible outside the finally block when the mutation occurs in a
         /// finally block.
         /// </summary>
         private void SetStateAndTrackForFinally(ref LocalState state, int slot, NullableFlowState newState)
         {
             state[slot] = newState;
-            if (newState == NullableFlowState.MaybeNull && NonMonotonicState.HasValue)
+            if (newState != NullableFlowState.NotNull && NonMonotonicState.HasValue)
             {
                 var tryState = NonMonotonicState.Value;
-                tryState[slot] = NullableFlowState.MaybeNull;
+                tryState[slot] = newState;
```

Comment:
>newState [](start = 33, length = 8)

This is not correct.  I believe `newState` on this line should be `newState.Meet(tryState[slot])`